pkg/distributor/distributor.go

```diff
@@ -338,128 +370,184 @@ func (d *Distributor) Push(ctx context.Context, req *client.WriteRequest) (*clie
+	// Obtain a subring if required
+	if size := d.limits.SubringSize(userID); size > 0 {
+		h := client.HashAdd32(client.HashNew32(), userID)
+		subRing, err = d.ingestersRing.Subring(h, size)
+		if err != nil {
+			return nil, httpgrpc.Errorf(http.StatusInternalServerError, "unable to create subring: %v", err)
+		}
+	}
+
+	if len(seriesKeys) == 0 && len(metadataKeys) == 0 {
+		// Ensure the request slice is reused if there's no series or metadata passing the validation.
 		client.ReuseSlice(req.Timeseries)
 
 		return &client.WriteResponse{}, firstPartialErr
 	}
 
 	now := time.Now()
-	if !d.ingestionRateLimiter.AllowN(now, userID, validatedSamples) {
+	//TODO: might be a problem for customers over the tipping point. Maybe increase limits by around 5 - 10%?
+	totalN := validatedSamples + len(validatedMetadata)
+	if !d.ingestionRateLimiter.AllowN(now, userID, totalN) {
 		// Ensure the request slice is reused if the request is rate limited.
 		client.ReuseSlice(req.Timeseries)
 
 		// Return a 4xx here to have the client discard the data and not retry. If a client
 		// is sending too much data consistently we will unlikely ever catch up otherwise.
 		validation.DiscardedSamples.WithLabelValues(validation.RateLimited, userID).Add(float64(validatedSamples))
-		return nil, httpgrpc.Errorf(http.StatusTooManyRequests, "ingestion rate limit (%v) exceeded while adding %d samples", d.ingestionRateLimiter.Limit(now, userID), numSamples)
+		validation.DiscardedMetadata.WithLabelValues(validation.RateLimited, userID).Add(float64(len(validatedMetadata)))
+		return nil, httpgrpc.Errorf(http.StatusTooManyRequests, "ingestion rate limit (%v) exceeded while adding %d samples and %d metadata", d.ingestionRateLimiter.Limit(now, userID), numSamples, len(req.Metadata))
 	}
 
-	var subRing ring.ReadRing
-	subRing = d.ingestersRing
+	if len(metadataKeys) > 0 {
+		err = ring.DoBatch(ctx, subRing, metadataKeys, func(ingester ring.IngesterDesc, indexes []int) error {
+			metadata := make([]*client.MetricMetadata, 0, len(indexes))
+			for _, i := range indexes {
+				metadata = append(metadata, validatedMetadata[i])
+			}
 
-	// Obtain a subring if required
-	if size := d.limits.SubringSize(userID); size > 0 {
-		h := client.HashAdd32(client.HashNew32(), userID)
-		subRing, err = d.ingestersRing.Subring(h, size)
+			localCtx, cancel := context.WithTimeout(context.Background(), d.cfg.RemoteTimeout)
+			defer cancel()
+			localCtx = user.InjectUserID(localCtx, userID)
+
+			if sp := opentracing.SpanFromContext(ctx); sp != nil {
+				localCtx = opentracing.ContextWithSpan(localCtx, sp)
+			}
+
+			return d.sendMetadata(localCtx, ingester, metadata)
+		}, func() {})
 		if err != nil {
-			return nil, httpgrpc.Errorf(http.StatusInternalServerError, "unable to create subring: %v", err)
+			// Metadata is a best-effort approach so we consider failures non-fatal, log them, and move on.
+			logger := util.WithContext(ctx, util.Logger)
+			level.Error(logger).Log("err", err)
```

Comment:
> If it's true that a push request can only contain series or metadata, why don't we return error? 

It is for now, but more importantly, it is something that we think will change in the future (given other Prometheus integrations e.g. Stackdriver) require metadata to be propagated with the series. The key takeaways here are a) We don't want to couple that implementation with Cortex and b) We want to prioritise samples over metadata.

I think we can this conversation again when I submit the refactor to send series + samples together to ingesters as the problem will be pushed there, but for now, I would feel more confident to just log the error for now and make sure we ingester the samples. WDYT?

Reply:
In the code above there's the following comment. I think it's misleading. Either we assume a request can contain both or not. I'm OK if we do, but we should simply reflect that comment just saying `A WriteRequest can include samples and metadata`.

```
	// A WriteRequest can now include samples and metadata. However, at the moment of writing,
	// we can only receive samples and no metadata or metadata and no samples.
```

About whether returning error or not, I'm OK keeping it as is (just logging error). We can always change it in the future to more stricter policy if required.

Reply:
> In the code above there's the following comment. I think it's misleading. Either we assume a request can contain both or not.

It's a bit tricky, but I see your point. I'll change to state:

> A WriteRequest can only contain series or metadata but not both. This might change in the future.